pkg/reduce/reduce.go

```diff
@@ -82,19 +82,25 @@ func NewDataForward(ctx context.Context,
 // Start starts forwarding messages to readloop
 func (d *DataForward) Start(ctx context.Context) {
 	d.readloop.Startup(ctx)
+	rCtx, rCancel := context.WithCancel(ctx)
 	for {
 		select {
 		case <-ctx.Done():
 			d.log.Infow("Stopping reduce data forwarder...")
 			cctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
 			defer cancel()
-			d.readloop.ShutDown(cctx)
 
 			if err := d.fromBuffer.Close(); err != nil {
 				d.log.Errorw("Failed to close buffer reader, shutdown anyways...", zap.Error(err))
 			} else {
 				d.log.Infow("Closed buffer reader", zap.String("bufferFrom", d.fromBuffer.GetName()))
 			}
+
+			// cancel the readloop context so that all the messages are persisted and acked.
+			rCancel()
```

Comment:
When `ctx` gets SIGTERM, the `rCtx` also has SIGTERM, which means, which means using explicit `rCancel()` is not needed.